hqdlock.hpp

```diff
@@ -1,427 +1,433 @@
+	 */
+	enum class pinning_policy_t {free_threads, pinned_threads};
+
+	/**
+	 * @brief hierarchical queue delegation lock implementation
+	 * @tparam GMLock global mutual exclusion lock
+	 * @tparam NMLock node-local mutual exclusion lock
+	 * @tparam DQueue delegation queue
+	 */
+	template<class GMLock, class NMLock, class DQueue, pinning_policy_t pinning_policy=pinning_policy_t::free_threads, starvation_policy_t starvation_policy=starvation_policy_t::starvation_free>
+	class hqdlock_impl {
+		template<class, class> friend class qd_condition_variable_impl;
+		
+		typedef qdlock_base<NMLock, DQueue, starvation_policy> base;
+		typedef hqdlock_impl<GMLock, NMLock, DQueue, pinning_policy, starvation_policy>* this_t;
+		struct hierarchy_sync {
+			static void lock(base* t) {
+				reinterpret_cast<this_t>(t->__data)->global_lock.lock();
+			}
+			static void unlock(base* t) {
+				reinterpret_cast<this_t>(t->__data)->global_lock.unlock();
+			}
 		};
-	};
-	typedef hierarchy_sync hierarchy_t;
-	typedef typename base::no_reader_sync reader_indicator_t;
-	
-	GMLock global_lock;
-	char pad1[128];
-	int numanodes;
-	std::vector<int> numa_mapping;
-	base* qdlocks;
-	thread_local static bool node_id_init;
-	thread_local static int node_id;
-	base& get_qd() {
-		int idx;
-		if(pinning_policy == pinning_policy_t::free_threads) {
-			int cpu = sched_getcpu();
-			idx = numa_mapping[cpu];
-		} else {
-			if(!node_id_init) {
-				node_id = sched_getcpu();
-				node_id_init = true;
-			}
-			idx = numa_mapping[node_id];
+		typedef hierarchy_sync hierarchy_t;
+		typedef typename base::no_reader_sync reader_indicator_t;
+		
+		GMLock global_lock;
+		char pad1[128];
+		int numanodes;
+		std::vector<int> numa_mapping;
+		base* qdlocks;
+		thread_local static bool node_id_init;
+		thread_local static int node_id;
+		base& get_qd() {
+			int idx;
+			if(pinning_policy == pinning_policy_t::free_threads) {
+				int cpu = sched_getcpu();
+				idx = numa_mapping[cpu];
+			} else {
+				if(!node_id_init) {
+					node_id = sched_getcpu();
+					node_id_init = true;
+				}
+				idx = numa_mapping[node_id];
+			}
+			return qdlocks[idx];
 		}
-		return qdlocks[idx];
-	}
-	public:
+		public:
 
-		hqdlock_impl() {
-			int num_cpus = sysconf(_SC_NPROCESSORS_CONF); // sane default
-			numa_mapping.resize(num_cpus, 0);
-			#ifdef QD_USE_LIBNUMA
-			/* use libnuma only if it is actually available */
-			if(numa_available() != -1) {
-				numanodes = numa_num_configured_nodes();
-				/* Initialize the NUMA map */
-				for (int i = 0; i < num_cpus; ++i) {
-					numa_mapping[i] = numa_node_of_cpu(i);
+			hqdlock_impl() {
+				int num_cpus = sysconf(_SC_NPROCESSORS_CONF); // sane default
+				numa_mapping.resize(num_cpus, 0);
+				#ifdef QD_USE_LIBNUMA
+				/* use libnuma only if it is actually available */
+				if(numa_available() != -1) {
+					numanodes = numa_num_configured_nodes();
+					/* Initialize the NUMA map */
+					for (int i = 0; i < num_cpus; ++i) {
```

Comment:
Since we are re-writing this code, wouldn't be a better idea to use `auto` (or `size_t`) instead of `int` in these loops (and in similar ones below)?

Also, `sysconf()` returns a `long`, not an `int`.  Perhaps you want to do something about this.